add dnnl patterns for byoc, along with their test units

crazydemo opened the pull request:
The test unit structure is inspired by `test_tensorrt.py`.
I consider this PR will be the next PR that is to be submitted to `apahce/tvm/master`.

gromero committed:
[TVMC] run: Don't use static path to find model.tar (#9712)

* [TVMC] run: Don't use static path to find model.tar

Currently 'tvmc run' when '--device micro' is specified looks for the
model in the project directory at <project_dir>/model.tar. That works
for Zephyr but fails on Arduino because model.tar is actually located at
<project_dir>/src/model/model.tar. As a consequence 'tvmc run' when used
to run a model on Arduino exists because model.tar is never found.

This commit fixes it by using the MLF path returned by the Project API
instead of using a static path.

This commit also adds a project_dir attribute to TVMCPackage that can be
set when a MLF archive is loaded/imported so the project dir can be
conveniently found (similarly to package_path attribute).

Signed-off-by: Gustavo Romero <gustavo.romero@linaro.org>

* [TVMC] test: Add test for importing a MLF with project_dir

Add test for TVMCPackage when importing a MLF archive and setting a
project directory too. Setting a project dir is only supported when a
MLF model is imported, so it must fail on Classic format.

Signed-off-by: Gustavo Romero <gustavo.romero@linaro.org>

asparkhi committed:
Follow up from CMSIS-NN pooling failure (#9708)

This commit fixes few comments in TIR2Runtime pass of CMSIS-NN target.
These comments specify layout used by CMSIS-NN API 
for input and filter shapes.

Another fix was done to the filter layout calculations. 
Instead of hard coded values for dimensions, 
filter_shape.find("H") was used to locate a particular value.

Third fix was done to the padding API used by Conv2D and Pooling tests.
It was made generic for TFLu's "SAME" padding type.

ophirfrish committed:
Improve tvmc error message from lazy-loading frontend imports (#9074)

When installing TVM from the python package, the Frontend frameworks dependencies such as TensorFlow, PyTorch, ONNX, etc, are not installed by default.
In case a user tries to run tvmc using a model whose framework was not installed, it will be presented with a very raw Python exception in the output.
The aim of this commit is to implement a better error messages for errors related to lazy-loading frontend frameworks in tvmc.

Change-Id: Ida52fac4116af392ee436390e14ea02c7090cef0

Commit:
add unit test for byoc-dnnl

michalpiszczek committed:
[TOPI] Add generic batch norm (#9694)

* Add topi batch norm and tests

* Handle none values correctly

* Return correct nun outputs for onnx

* Use moving var/mean and update tests

* Add a test for batch norm folding

* Fix comment

* Format with black

* Re-order test args to match interface

* Call fold constant manually

leeexyz committed:
[TVMC] Add --opt-level to compile mode (#9722)

hzfan committed:
[Relay] Non-recursive dependency graph (#9528)

* fix recur

* fix ci

driazati committed:
Add `make docs` and doc building instructions (#9534)

* Update doc building instructions and pin dependencies

This pins the dependencies for the docs and adds `pytest` as a
dependency which was missing when I built. Tested out the
requirements.txt with a fresh `ubuntu:focal` Docker image to verify that
the required depedencies work.

* Address comments, add Makefile for docs and add to instructions

* Use Python for running scripts

* Fix lint, add lint command

* Add option for cpu to only run the precheck, address comments

* Fix 'make doc' usage, add some -x's

* Fix bad condition on --cpu, add defaults for envs

* Fix another 'make doc'

* Fix for running on MacOS

Co-authored-by: driazati <driazati@users.noreply.github.com>

gromero committed:
Fix typos in runtime comments (#9726)

Fix typos in comments about signal handling and entries in the runtime
code.

Signed-off-by: Gustavo Romero <gustavo.romero@linaro.org>

MasterJH5574 committed:
[BugFix] Fix a wrong use of `std::move()` in cross-thread reduction lowering (#9728)

* [BugFix] Fix a wrong use of `std::move()` in cross-thread reduction lowering

* Remove

masahi committed:
[CUTLASS] More robust support for pattern matching and alignment (#9698)

* bug fix in im2col encoding

* skip legalize when batch size is dynamic

* add sm75 kernels to sm80 profilings

* add dtype and layout check in parttern match

* use align1 kernel for unusual channel cases (IC = 3 etc)

* test IC=3 convolution

* fixed check functions for fused cases, run infer type before mergecomposite

* check align on N dim

* add comment on IC == 3 case

* lint fix

* do not offload depthwise conv2d

* lint

* trigger CI

masahi committed:
[Relay, BYOC] Make constant binding in PartitionGraph optional (#9721)

* make constant binding in PartitionGraph optional

* add test

* Doc string update

* remove default argument from set_body_typed

driazati committed:
Add CMake summary (#9696)

* Add CMake summary

This adds a printout at the end of CMake's configuration that prints out all the user options (anything declared via `tvm_option`) plus some extra metadata about CMake itself (version, system, etc). This makes it easier to see what CMake is doing without guessing from config files (i.e. is the build release or debug mode, is CUDA on, ...)

* Add SUMMARIZE=ON to CI cmake configs

Co-authored-by: driazati <driazati@users.noreply.github.com>

Commit:
add byoc-dnnl pattern and their test cases

masahi committed:
[CUDA] Do not emit vector load on unaligned base offset (#9731)

* [CUDA] Do not emit vector load on unaligned base offset

* fix alignment check condition

* black

* improve test

* improve the vectorization condtion to avoid error in yolo5 (thanks to vinx13)

* replace coeff != 1 check by coeff % lane == 0

driazati committed:
Merge Java unittests into GPU unittests (#9732)

These run on the same node but the Java unit tests take < 1 min (whereas the node setup alone takes 3-4 min), so it seems like it would be good to just run the Java tests as part of the general GPU unit testing rather than allocate and set up a separate node for it.

Co-authored-by: driazati <driazati@users.noreply.github.com>

grant-arm committed:
[microNPU] Update Arm(R) Ethos(TM)-U55 NPU demo README (#9725)

- Update README.md to show config.cmake options when building from source

Change-Id: Iecc36d8bc889541b8b890de37d58052a46928a83

mbs-octoml committed:
[Relay] Re-run PlanDevices after LowerTE to flow new memory scope constraints. (#9613)

* [Relay] Re-run PlanDevices after LowerTE to flow new memory scope constraints.

This PR:
 1) Makes PlanDevices consider lowered calls when solving device domain constraints.
 2) Connects the storage scopes on PrimFunc parameters (encoded in their Buffer data
    Var type annotation PointerTypes storage_scope fields) to the memory_scope
    fields of the SEScopes which PlanDevices unifies over.
 3) Allows new device_copies to be inserted on the arguments and results of lowered
    calls so as to acount for any memory scope mismatches which are now apparent.

[device_planner.cc has main changes, rest is secondary.]

In the short term we'd like to use this machinery to flow memory scope choices made
during lowering back out into the overall Relay program. In the longer term we'd
also like to be able to use memory scopes to influence the lowering of
yet-to-be-lowered functions (or lowered functions which have yet to been scheduled,
a distinction now possible with TensorIR).

 - Memory scope constraints can flow both out of and in to PrimFuncs
   introduced by LowerTE. In TIR memory scopes are represented by
   'storage scopes' on the PointerType type annotations on TIR Buffer data
   variables.
    - It is straightforward to extract memory scopes from PrimFuncs by
      looking at the PrimFunc's buffer_map. We do this is 'phase 1' of
      PlanDevices, which collects all the device constraints implied by
    - However, pushing memory constraints in to PrimFuncs is more challenging
      due to buffer aliasing. This aspect is still experimental.

 - Allow device_copies to be inserted for both arguments and
   results of PrimFunc calls, on the assumption PlanDevices has
   already established a consistent device assignment prior to
   lowering and any new mismatch is required to match up memory scopes.
   We use the new 'free' on_device annotations to implement this.

Coming along for the ride:

 - To make unit tests of mixed Relay/TIR functions possible needed
   to be able to supply a checked_type to GlobalVar since that's currently
   the only way to give a Relay type to PrimFuncs.

 - Use GenSym to get unique var names in ANF & partial eval so easier
   to diff debug output between passes and connect program fragments
   back into the overall program. Relying on pretty-printing to
   automagically unique-ify var names is certainly cute but until we
   have better span support is very hard to work with.

 - Realized both dead_code.cc and fold_constant.cc would
   happily move values into a different lexical virtual
   device context since device_planner.cc was being
   'clever' and eliding on_devices for let-bound values
   when there's no change. Fixed so that every let-bound
   value has an on_device. Will be much better after
   https://github.com/apache/tvm-rfcs/pull/45 is implemented.

 - Make build -Werror clean for clang-12 (mostly move fixups).

 - Address post-submit comments from #9693.

* [checkpoint] thread safe GenSym

tqchen committed:
[CI] Hotfix Jenkinsfile (#9739)

wrongtest-intellif committed:
Fix for tvm.build()'s name warning (#9678)

* change default name of tvm.build() to None

* change lower name default to None

* fix lint

* lower() use "main" -> "default_function"

* remove default name's warning

masahi committed:
[CUTLASS] Add conv2d profiler (#9737)

* Add cutlass conv2d profiler

commit 1c0bbb297da43dab75ad995afbcacd59e9fe4c87
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 18:29:03 2021 +0900

    fix lint

commit 463574ce087ca0444b23d4f47baf8066b8fbd3df
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 17:28:38 2021 +0900

    fixed conv2d check

commit 588c5abe15abbf0339a7972d81b8235bc7460620
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 15:05:27 2021 +0900

    update test

commit a447b57ade7c99da4efd38e1bdfc213a64a80fd2
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 14:54:52 2021 +0900

    speed up profiling by removing initialization

commit 93cd039ba04dd80e887ec1a71f358cd86a1e5221
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 08:26:29 2021 +0900

    fixed nhwc cudnn depthwise conv

commit 6db71727f553ee2009c9d3feb2b019c24459f4d2
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sat Dec 11 15:39:05 2021 +0900

    add cache

commit f7d17a116acd80c57dfa04aa86577fa30898908d
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sat Dec 11 15:05:38 2021 +0900

    removed im2col profiling for conv2d

commit b724f446d07030e45f30474a1e3c124f1541b496
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 22:57:54 2021 +0900

    black

commit fe4687b9f6d41eff66742cdde694680538a3d5e4
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 22:49:13 2021 +0900

    fixed cmd arguement

commit ab114f5c3e1a3086255caccd6c0f5df09d7c3755
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 22:22:19 2021 +0900

    conv2d profiler working

commit 49ee61f5583f73f0d9b2c18c1861c90fa028f64a
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 20:26:15 2021 +0900

    add conv2d profiler

commit 49e2c8918a1a2632f60c700675f12f9d83adef24
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 08:03:36 2021 +0900

    do not offload depthwise conv2d

commit cd8367768229720020d81d597d751bfa95ec014e
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 13:20:01 2021 +0900

    lint fix

commit 870823c6d54114896aa9db91eb72c132cdef762f
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:54:38 2021 +0900

    add comment on IC == 3 case

commit 6b780db7f8059a9a58bfc257fbb9cf7cb60b72a2
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:48:33 2021 +0900

    check align on N dim

commit 308c4dac39f761ac157f032b9fe7028820980294
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:34:42 2021 +0900

    fixed check functions for fused cases, run infer type before mergecomposite

commit 8d6a1bfee26e9dfdeb1ab001aa89b43b9cd33d74
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:10:59 2021 +0900

    test IC=3 convolution

commit ffce47de724398222f672b220156b19c8f48a700
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:10:16 2021 +0900

    use align1 kernel for unusual channel cases (IC = 3 etc)

commit 6cdf205a3451b5fa7ea0cbff0228ae1da775573a
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:06:56 2021 +0900

    add dtype and layout check in parttern match

commit 7743cc6dde442b45d66b4be320268ed24f352422
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 10:40:53 2021 +0900

    add sm75 kernels to sm80 profilings

commit efceccb994ec71bc4ba847c791fc4f696eb7f242
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 10:40:42 2021 +0900

    skip legalize when batch size is dynamic

commit 65fbc0a0813cb4e980e53240f34f3ba18754ab99
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 10:36:36 2021 +0900

    bug fix in im2col encoding

* minor fix

* lint fix

* allow autotvm NCHW depthwise conv2d schedule even if -libs=cudnn

* Update python/tvm/contrib/cutlass/gen_conv2d.py

Co-authored-by: Cody Yu <comaniac0422@gmail.com>

* simplify processing profiler outputs

* more simplify

* fix runtime check

Co-authored-by: Cody Yu <comaniac0422@gmail.com>

vinx13 committed:
[TIR] Add 'global_symbol' and 'tir.noalias' as default attributes in script auto completion (#9744)

* [TVMScript] Update default TIR prefix to T

* [TIR] Add 'global_symbol' and 'tir.noalias' as default attributes in script auto completion

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

vinx13 committed:
[TIR][Schedule] Analysis functions to check if compute_inline and com… (#9743)

* [TIR][Schedule] Analysis functions to check if compute_inline and compute_inline is allowed

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

* Address comments

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

mbs-octoml committed:
[Relay] Support large constants saved/loaded outside of VM executable (#9734)

* [Relay] Support large constants.

This allows constant tensors at or above a given byte limit to be marked as
'late bound' and saved/reloaded to a file independently of the overall
executable. Since the executable is often embedded in the data segment of
generated runtime Modules this avoids problems with external tools which can't
handle multi-gigabyte data segments.

[ACE-466 in OctoML JIRA]

* [checkpoint] fix latent bytecode/code bug

vinx13 committed:
Fix typo (#9740)

jiangjiajun committed:
[Frontend][PaddlePaddle] Enhance paddlepaddle frontend with more operators (#9724)

* add operators for paddle frontend

* add operators for paddle frontend

* retrigger ci

* retrigger ci

* retrigger ci

Co-authored-by: wjj19950828 <wjjisloser@163.com>

Commit:
fix compact_dataflow (#9747)

Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

MasterJH5574 committed:
[TensorIR] Primitive "SetScope" (#9738)

* Main code

* Reorder steps

* Unittests

* Docstring

* Check the input storage scope

* Docstring for `CheckStorageScope`

* Import header

adstraw committed:
Add unit tests for HexagonBuffer (#9736)

* Add unit tests for HexagonBuffer

* fix build error for signed / unsigned comparison

mbs-octoml committed:
[Relay] Add a unit test for structural equality (#9745)

This is CORE-135 from the forums, which suggested structural equality
was deeply broken. But unable to repro. No harm including unit test.

(attempt 3)

yangulei commented:
@Zhennan

yangulei commented:
@ZhennanQin

ciyongch commented on python/tvm/relay/op/contrib/dnnl.py:
```diff
@@ -76,12 +78,79 @@ def make_pattern(with_bias=True):
+    )
+    conv2d_bias_pat = ("dnnl.conv2d_bias", make_conv_pattern(with_bias=True, with_relu=False))
+    conv2d_relu_pat = ("dnnl.conv2d_relu", make_conv_pattern(with_bias=False, with_relu=True))
+    dense_bias_pat = ("dnnl.dense_bias", make_dense_pattern(with_bias=True, with_relu=False))
```
Is it easy to extend the pattern list for various elementwise post-ops other than relu (for example, tanh/sigmoid/mul, etc.)?

crazydemo replied:
It is easy to extend on dnnl codegen, but can be a bit hard on tvm, as it need to registster all ops first. 
I will try to figure out.

crazydemo replied:
I have updated the related code. 
If we want to extend the pattern list, we need to extend the `elt_list`.
And we also need to add pattern in `dnnl_json_runtime.cc`

ciyongch commented on python/tvm/relay/op/contrib/dnnl.py:
```diff
@@ -76,12 +78,79 @@ def make_pattern(with_bias=True):
+
+
+def partition_for_dnnl(
+    mod,
```
format? Keep the two args in the same line of function declaration.
I'm wondering if there's any type check for python code in TVM?

crazydemo replied:
fixed. TVM has C++ and python lint, all lint has been passes.

ciyongch commented on python/tvm/relay/op/contrib/dnnl.py:
```diff
@@ -76,12 +78,79 @@ def make_pattern(with_bias=True):
+            transform.PartitionGraph(),
+        ]
+    )
+    with tvm.transform.PassContext(opt_level=3):
```
what's the difference result when the `opt_level` equals to values like 0/1/2? do you think user need this option when creating partition for dnnl backend?

crazydemo replied:
if `opt_level=1 / 2`, some passes like alter_op_layout will be disable, then it cannot pass the test cases, `opt_level=3` is the default setting.

masahi committed:
[CUTLASS] Support conv2d activation fusion (#9746)

* Add cutlass conv2d activation (bias, relu, sigmoid)

commit e4e273ae74a8e54ab1ae1414ce9b6bfcc2b3d530
Merge: 0489d1418 77c938550
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Mon Dec 13 11:58:54 2021 +0900

    Merge branch 'partition-constant-unbind' into cutlass-conv2d-fusion

commit 77c9385501595f804bd33b436aed0cc192059a10
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Mon Dec 13 11:58:18 2021 +0900

    add test

commit ab01b3aae36ef88ec299ff5e9d1bbdf5fd7268f6
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Mon Dec 13 11:55:06 2021 +0900

    make constant binding in PartitionGraph optional

commit 0489d1418c5f49b95ad220b8f0e57ba48443c6a3
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 21:52:29 2021 +0900

    support sigmoid fusion (only fp32 accum for now)

commit 3705bbd6b77ec8d8c416ab00738fe8b9c91e1535
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 20:50:58 2021 +0900

    conv2d fusion test worked

commit 05b51c9f94c9a4f31afeb126c8749675be423892
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 20:34:10 2021 +0900

    fix bias stride

commit 7cf40e719a8c1c464e2ee925018d8ed90b145dcd
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 20:01:21 2021 +0900

    use nobetascaling

commit 274ec028845e296b012c23830a0be6f48cbee767
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 19:12:58 2021 +0900

    adding fusion support to codegen

commit 0de5ebdb2e318129a1a4d3a2e568849a993525c4
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 18:39:08 2021 +0900

    partition working

commit c08bb38e3eea90168dc036cb4cbfbcdef288bac6
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 17:24:42 2021 +0900

    update test

commit 81bf9e600b1bbdbc27e0e8b54496fc09b33e0624
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 13:23:39 2021 +0900

    add fused conv2d pattern

commit 1c0bbb297da43dab75ad995afbcacd59e9fe4c87
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 18:29:03 2021 +0900

    fix lint

commit 463574ce087ca0444b23d4f47baf8066b8fbd3df
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 17:28:38 2021 +0900

    fixed conv2d check

commit 588c5abe15abbf0339a7972d81b8235bc7460620
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 15:05:27 2021 +0900

    update test

commit a447b57ade7c99da4efd38e1bdfc213a64a80fd2
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 14:54:52 2021 +0900

    speed up profiling by removing initialization

commit 93cd039ba04dd80e887ec1a71f358cd86a1e5221
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 08:26:29 2021 +0900

    fixed nhwc cudnn depthwise conv

commit 6db71727f553ee2009c9d3feb2b019c24459f4d2
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sat Dec 11 15:39:05 2021 +0900

    add cache

commit f7d17a116acd80c57dfa04aa86577fa30898908d
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sat Dec 11 15:05:38 2021 +0900

    removed im2col profiling for conv2d

commit b724f446d07030e45f30474a1e3c124f1541b496
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 22:57:54 2021 +0900

    black

commit fe4687b9f6d41eff66742cdde694680538a3d5e4
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 22:49:13 2021 +0900

    fixed cmd arguement

commit ab114f5c3e1a3086255caccd6c0f5df09d7c3755
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 22:22:19 2021 +0900

    conv2d profiler working

commit 49ee61f5583f73f0d9b2c18c1861c90fa028f64a
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 20:26:15 2021 +0900

    add conv2d profiler

commit 49e2c8918a1a2632f60c700675f12f9d83adef24
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Sun Dec 12 08:03:36 2021 +0900

    do not offload depthwise conv2d

commit cd8367768229720020d81d597d751bfa95ec014e
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 13:20:01 2021 +0900

    lint fix

commit 870823c6d54114896aa9db91eb72c132cdef762f
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:54:38 2021 +0900

    add comment on IC == 3 case

commit 6b780db7f8059a9a58bfc257fbb9cf7cb60b72a2
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:48:33 2021 +0900

    check align on N dim

commit 308c4dac39f761ac157f032b9fe7028820980294
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:34:42 2021 +0900

    fixed check functions for fused cases, run infer type before mergecomposite

commit 8d6a1bfee26e9dfdeb1ab001aa89b43b9cd33d74
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:10:59 2021 +0900

    test IC=3 convolution

commit ffce47de724398222f672b220156b19c8f48a700
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:10:16 2021 +0900

    use align1 kernel for unusual channel cases (IC = 3 etc)

commit 6cdf205a3451b5fa7ea0cbff0228ae1da775573a
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 12:06:56 2021 +0900

    add dtype and layout check in parttern match

commit 7743cc6dde442b45d66b4be320268ed24f352422
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 10:40:53 2021 +0900

    add sm75 kernels to sm80 profilings

commit efceccb994ec71bc4ba847c791fc4f696eb7f242
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 10:40:42 2021 +0900

    skip legalize when batch size is dynamic

commit 65fbc0a0813cb4e980e53240f34f3ba18754ab99
Author: Masahiro Masuda <masahi129@gmail.com>
Date:   Fri Dec 10 10:36:36 2021 +0900

    bug fix in im2col encoding

* support batch norm fusion

vinx13 committed:
[TIR][Schedule] Add Annotate/Unannotate primitive (#9742)

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Xiyou Zhou <xiyou@octoml.ai>

junrushao committed:
[MetaSchedule] Add the missing HasWorkload interface to the Database (#9756)

Commit:
[TVMScript][FIX] Fix number of arguments for T.Buffer[...] (#9758)

* fix number of arguments

* make test clear

* Update tests/python/unittest/test_tvmscript_syntax_sugar.py

Co-authored-by: Wuwei Lin <vincentl13x@gmail.com>

* only tuple for now

Co-authored-by: Wuwei Lin <vincentl13x@gmail.com>

Commit:
Move Compute library to 21.11 (#9754)

mbs-octoml committed:
[Relay] s/SEScope/VirtualDevice/g (#9759)

* [Relay] s/SEScope/VirtualDevice/g

Nobody liked 'SEScope', and 'DeviceMcDeviceFace' is too verbose, so it
seems 'VirtualDevice' has the popular vote.

huajsj committed:
[Runtime][Pipeline Executor] Add the map logic of global input and subgraph input. (#9751)

* [Runtime][Pipeline Executor] Add the map logic of global input and subgraph input.
User can use "global input name" to feed input data for pipeline runtime. The name like
"data_a" will be mapped into a input interface of subgraph. In this PR, we
create the related logic to do the following things.
 1. building the input map configuration
 2. in runtime c++ module, parseing the input connection configuration then
    creating related data structure to record the said connection map.
 3. providing the function to return the map information for verification.

* address review comments.

* addres review comments.

* address review comments.

junrushao committed:
[MetaSchedule] Random Feature Extractor (#9760)

Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

mbs-octoml committed:
[Relay] Fix invalid shape function for "copy" operator (#9749)

The 'script' for of the shape function was ill-formed,
resulting in a TIR shape function which did not assign
to it's output, which in turn caused either OOM or
assert fails as uninitialized dimensions worked their
way downstream. That fix is in python/tvm/relay/op/tensor.py.

Everything else is for testing and debugging as I tracked
this down.

Special thanks to Lily for helping me with the scalar vs
tensor switch in the copy shape function.

[This is CORE-112 in OctoML JIRA.]

driazati committed:
Remove CMake string REPEAT (#9771)

According to https://cmake.org/cmake/help/latest/command/string.html#repeat this was added in 3.15, we can revert this if we ever use a minimum cmake newer than that. Until then we should just remove the alternate path to reduce complexity.

Co-authored-by: driazati <driazati@users.noreply.github.com>

sunggg committed:
[MetaSchedule] Misc improvement of the Measurer (#9757)

Co-authored-by: Junru Shao <junrushao1994@gmail.com>

zxybazh committed:
[M3c][MetaScheduler] Add ScheduleRule class & PostOrderApply space generator. (#9761)

* Add ScheduleRule class & PostOrderApply space generator.

Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

* Fix comments & docs.

* Fix for mypy.

* Retrigger CI.

* remove get_hex_address

Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

zhiics committed:
[community] @lunderberg -> Committer (#9773)

driazati committed:
Clarify error message for missing libraries (#9710)

Ran into this when importing tvm without a build

Co-authored-by: driazati <driazati@users.noreply.github.com>

Commit:
[TE][TensorIR] fix tensor attr in create_prim_func (#9764)

* [TE][TensorIR] fix tensor attr in create_prim_func

* Update src/te/operation/create_primfunc.cc

Co-authored-by: Junru Shao <junrushao1994@gmail.com>

Co-authored-by: Junru Shao <junrushao1994@gmail.com>

pansn1995 committed:
Fix GLOBAL_SCOPE Shallow copy bug (#9718)

wrongtest-intellif committed:
[TIRScript] fix parse StringImm value in for loop annotations (#9755)

* fix parse strimm value in for annotations

* flatten buffer allow runtime.String attr value

* remove unused import

* rebase and ensure flattened attr order

gussmith23 committed:
Include \0 terminating character in strncpy (#9775)

I'm fairly certain that this bugfix is correct, but please check my code. I'm not sure how this hasn't caused more problems for people! Perhaps the malloced memory is usually filled with 0s? Either way, on my system, the malloced memory was NOT all 0s, and so when we would copy the `key.size()` valid characters of `key` into `*out_type_key`, it was likely that the next character was not 0, and thus, because no terminator was copied over, the string wouldn't be terminated correctly. This fix ensures the terminator gets copied over, by copying `key.size() + 1` characters.

Commit:
fix int set analysis on negative scale (#9776)

jacobbohlin committed:
[microNPU] Added checks for out of range shifts (#9707)

* [microNPU] Added checks for out of range shifts

* Added testcase

* Addressed comments

Commit:
[Hexagon] Account for objects being smaller than the allocated space (#9769)

* [Hexagon] Account for objects being smaller than the allocated space

In particular, graph executor will reuse allocated buffers for various
tensors. These tensors may be of various sizes as long as the buffer
is large enough to hold them.

* Trigger CI

* Add #include <algortihm> to hexagon_buffer.cc

* Update cpptest, plus fix one CHECK in hexagon_buffer

* Rename `offset` to `copied` for consistency

jacobbohlin committed:
[microNPU] Fix incorrect comparison in schedulers (#9706)

* [microNPU] Fix incorrect comparison in planning functions

Added diamond graph test case to test_scheduler.py

* Fixed copy luts test case

grant-arm committed:
[microNPU] Upgrade to 21.11 version of Arm(R) Ethos(TM)-U55 NPU driver (#9777)

Change-Id: Ide4d2a33a215b4f1367667ad0fa4a66913cf9ad4

qimaolvshangxue committed:
[Bugfix]  fix the bug that occurs when the test_pass_ctx_exception() is (#9774)

tested separately.

In the batch test ,cause a previous function will specify the value of
PassContext.current() as None,so tests the test_pass_exception() will pass.
but when testing it individually ,the result of PassContext.current() is
a [] instead of None,it will not pass, this may happen when using the
option of "pytest -n 4"

ophirfrish committed:
Improve the frontend tflite _test_abs test to support tflite 2.6 (#9783)

Updated the test input node name

masahi committed:
[AMP] Disallow converting layer norm to fp16 (#9782)

* [AMP] Disallow converting layer norm to fp16

* black

AndrewZhaoLuo committed:
[Autoscheduler] Task Extraction Raises Exception on Lowering (#9750)

* forward messages

* Update python/tvm/auto_scheduler/relay_integration.py

Co-authored-by: Cody Yu <comaniac0422@gmail.com>

* remove type annotation for consistency

* lint

Co-authored-by: Cody Yu <comaniac0422@gmail.com>

ciyongch commented on python/tvm/relay/op/contrib/dnnl.py:
```diff
@@ -76,12 +81,83 @@ def make_pattern(with_bias=True):
+    elif op == "dense":
+        return (pat_name, make_dense_pattern(with_bias, with_eltwise))
+    else:
+        return None
```
suggest to add some warning msg like `Currently, only conv2d and dense op are supported, but got ()".format(op).`

crazydemo replied:
warning log has been added.

ciyongch commented on python/tvm/relay/op/contrib/dnnl.py:
```diff
@@ -76,12 +81,83 @@ def make_pattern(with_bias=True):
+    dnnl_patterns = []
+    for with_bias in [True, False]:
+        for elt in elt_list:
+            if not with_bias and not elt:
+                return dnnl_patterns
```
no need these without "None"  in elt_list.

crazydemo replied:
I need this to construct patterns like `conv-bias`, who has no activation attached.

ciyongch replied:
Ok, then it should be fine.

ciyongch reviewed (commented):
Overall looks good:)

wrongtest-intellif committed:
[CodeGen] avoid crash if an exception is raised during llvm cpu codegen (#9786)

* avoid crash if an exception is raised during llvm cpu codegen

* use pytest.raises

zxybazh committed:
[M3c][MetaScheduler] Add More Measure Callbacks. (#9780)

* Add measure callbacks.

Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

* Fix comments.

Co-authored-by: Junru Shao <junrushao1994@gmail.com>
Co-authored-by: Bohan Hou <32121147+spectrometerHBH@users.noreply.github.com>
Co-authored-by: Ruihang Lai <lairuihangdongdong@qq.com>
Co-authored-by: Hongyi Jin <3231950289@qq.com>
Co-authored-by: Wuwei Lin <wuwei@apache.org>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

ciyongch reviewed (approved):
LGTM.
